wagtail/core/collectors.py

```diff
@@ -0,0 +1,669 @@
+        #   object references, as specified by block_types above
+        filters = Q()
+        for collector, block_paths in block_paths_per_collector:
+            field_name = collector.field.attname
+
+            # for each block in the StreamField's definition that might contain an object
+            # reference, as specified by block_types above, find the deepest-nested named
+            # block in its path. (Not all blocks are named; the child block of a ListBlock
+            # isn't, for example.)
+            last_blocks = [[block for block in block_path if block.name][-1]
+                           for block_path in block_paths]
+            block_filter = Q(**{field_name + '__regex': r'"(%s)"'
+                                % '|'.join({block.name for block in last_blocks})})
+            value_filter = (
+                Q(**{field_name + '__regex': pattern})
+                if searched_objects else Q())
+            filters |= block_filter & value_filter
+
+        searched_data = {get_obj_base_key(v) for v in searched_objects}
+        for obj in self.model._default_manager.using(self.using) \
+                .filter(filters):
+            for collector, block_paths in block_paths_per_collector:
+                for block_path in block_paths:
+                    for found_value in collector.find_objects(
+                            getattr(obj, collector.field.attname), block_path):
+                        if get_obj_base_key(found_value) in searched_data:
+                            yield obj, found_value
+
+
+class Use:
+    def __init__(self, obj: Model, parent=None, on_delete=CASCADE, field=None):
+        self.object = obj
+        self.key = get_obj_base_key(obj)
+        self.parent = parent
+        self.on_delete = on_delete
+        self.field = field
+
+        self.depth = 0
+        while parent is not None:
+            self.depth += 1
+            parent = parent.parent
+
+    @classmethod
+    def from_flat_iterable(cls, iterable, on_delete=CASCADE, field=None,
+                           exclude=None):
+        if exclude is None:
+            exclude = set()
+        for obj in iterable:
+            use = cls(obj, on_delete=on_delete, field=field)
+            if use not in exclude:
+                exclude.add(use)
+                yield use
+
+    @classmethod
+    def from_nested_list(cls, nested_list, parent=None, on_delete=CASCADE,
+                         exclude=None, originals=()):
+        if exclude is None:
+            exclude = set()
+        main_use = None
+        for i, obj in enumerate(nested_list):
+            if isinstance(obj, list):
+                yield from cls.from_nested_list(
+                    obj, parent=main_use, on_delete=on_delete, exclude=exclude)
+            else:
+                use = cls(obj, parent=parent, on_delete=on_delete)
+                if use in exclude:
+                    if use not in originals:
```

Comment:
@BertrandBordage If multiple objects are being passed to `get_all_uses`, it's possible for one object to be in the `originals` list and _also_ part of a `CASCADE` chain for another of those objects, isn't it? If so, we can't just skip over this logic for objects in `originals`.

Reply:
...in fact, what's the purpose of this line at all? As far as I can see, the only reason why we _might_ want to give `originals` special treatment is that they exist in the `exclude` list as plain instances rather than `Use` records, and it may or may not be a good thing to populate `Use.parent` with a plain model instance rather than a `Use`.

Given that the `parent` attribute of `Use` doesn't appear to be used, tested or documented anywhere, I'm inclined to remove it completely.

Reply:
@gasman, to answer your questions (sorry in advance :slightly_frowning_face:):
- `originals` and `exclude` always contain `Use` instances, not `Model` instances.
- `Use.parent` is used in `Use.__init__`. Each `Use` instance uses the `parent` attribute of their ancestors to define its depth. I agree that it could be removed and increment `parent.depth` instead. However, I still think it’s a good thing to keep a `parent` attribute in a tree structure.
- That whole `if use in exclude: if use not in originals: […]` has a single purpose: work around the messy nested structure returned by Django’s `NestedObjects`. It can return something like this for an object `o1`: `[o1, [o5], o2, [o4], o3, [o1]]` (note that I’m simplifying because the example doesn’t contain multitable inheritance, which is correctly handled by this method). In that case:
  - `o2`, `o3` and `o5` are directly related to `o1`. Yes, it doesn’t seem to make sense because `o5` is nested while `o2` and `o3` are at the same level as `o1`. This is the main cause for the complexity of this method. `[o5]` is considered as a list of children of `o1` because it’s following `o1`. `o2` and `o3` are at a different level because they also contain related objects. I know, this is an awful tree structure, but that’s what we have to deal with.
  - `o4` is related to `o2`. We want to list it, but unlike `o5` we have to remember that `o4` is contained in `o2`, whereas `o5` was considered at the root of our uses tree structure. So this means we had to remember the previous item, `o2`, as the parent for the list of children, and therefore we store it in `main_use`. Before that, we search for `other_use`, the first instanciation of `o2` in case there is any other. This is to avoid using extra RAM because millions of uses can be found for a single deletion…
  - `o1` is related to `o3`, but we don’t want to list `o1` since we already know we want to delete it. Since `o1` is already in `exclude` from `get_all_uses`, it will not be shown. To link `o3` to `o1`, we defined the variable `main_use` containing the use preceding the list `[o1]`.
- It’s not possible for one object from `originals` to be in the `CASCADE` chain of another object. `NestedObjects` can only return an object once as a non-leaf node. For example, you couldn’t get `[o1, [o2, [o1, [o3]]]]`. There are two reasons for that, first, obviously to avoid infinite recursion, and second, because of the SQL optimisations done by `NestedObjects` to avoid triggering too many requests (even though it’s far from being optimised :frowning_face:).

**TLDR:**
**Without `if use not in originals:`** (wrong because it shouldn’t show a chevron because nothing is contained)
![image](https://user-images.githubusercontent.com/1119169/43279560-ab091b4c-910e-11e8-88d3-22f2b09cfbe7.png)
**With `if use not in originals:`** (right solution)
![image](https://user-images.githubusercontent.com/1119169/43279614-df083dec-910e-11e8-8b37-f66ae3d9cf97.png)

Reply:
That's fine, but can this go in code comments please? :-)

Also, could we perhaps remove the `Use(foo) == foo` behaviour from `Use.__eq__`? I noticed that the tests are relying on that behaviour, but I have no idea if any other code is - and it's yet another thing that's making this PR extremely difficult to review, because every time I see a comparison between `Use` objects, I have to stop and ask myself if it's doing the obvious thing or whether there's some secret type-hacking going on.

Reply:
Yeah, no problem with removing the ability to compare `Use` and `Model` instances.

Well, you see the problem with putting all that as code comments x)
Tricky to do such complex explanations while keeping the code readable. Do you want me to add a docstring+a few inline comments in this method? I can make a commit in my old branch that you could cherry-pick.

Reply:
Yes please!